teamcity/datadog_checks/teamcity/common.py

```diff
@@ -0,0 +1,152 @@
+    exclude_patterns = config_key.get('exclude', default_exclude)
+    filtered_items = filter_list(items, include_patterns, exclude_patterns)
+    reached_limit = len(OrderedDict(list(filtered_items.items())[0:limit])) < len(list(filtered_items.items()))
+    return OrderedDict(list(filtered_items.items())[0:limit]), reached_limit
+
+
+def filter_projects(check, projects_list):
+    return filter_items(
+        items=projects_list,
+        key='projects',
+        config=check.instance,
+        default_limit=check.default_projects_limit,
+        default_include=None,
+        default_exclude=None,
+    )
+
+
+def filter_build_configs(check, build_configs_list, project_pattern, config):
+    return filter_items(
+        items=build_configs_list,
+        key=project_pattern,
+        config=config,
+        default_limit=check.default_build_configs_limit,
+        default_include=check.global_build_configs_include,
+        default_exclude=check.global_build_configs_exclude,
+    )
+
+
+def construct_event(check, new_build):
+    """
+    Construct event based on build status.
+    """
+    build_id = new_build['id']
+    build_number = new_build['number']
+    build_config = new_build['buildTypeId']
+    instance_name = check.instance_name if check.instance_name else build_config
+    build_status = STATUS_MAP.get(new_build['status'])['msg_title']
+    alert_type = STATUS_MAP.get(new_build['status'])['alert_type']
+    event_tags = deepcopy(check.build_tags)
+    event_tags.extend(['build_id:{}'.format(build_id), 'build_number:{}'.format(build_number)])
+
+    if check.is_deployment:
+        teamcity_event = deepcopy(DEPLOYMENT_EVENT)
+        teamcity_event['tags'].append('type:deployment')
+    else:
+        teamcity_event = deepcopy(BUILD_EVENT)
+        teamcity_event['tags'].append('build')
+
+    teamcity_event['timestamp'] = int(time.time())
+    teamcity_event['msg_title'] = teamcity_event['msg_title'].format(
+        instance_name=instance_name, host=check.host, build_status=build_status
+    )
+    teamcity_event['msg_text'] = teamcity_event['msg_text'].format(
+        build_number=new_build['number'], host=check.host, build_webUrl=new_build['webUrl']
+    )
+    teamcity_event['alert_type'] = alert_type
+    teamcity_event['host'] = check.host
+    teamcity_event['tags'].extend(event_tags)
+
+    return teamcity_event
+
+
+def get_response(check, resource, **kwargs):
+    resource_url = RESOURCE_URL_MAP[resource].format(base_url=check.base_url, **kwargs)
+    resource_name = " ".join(resource.split("_"))
+
+    try:
+        resp = check.http.get(resource_url)
+        resp.raise_for_status()
+
+        json_payload = resp.json()
+        if resource == 'build_config' or resource == 'teamcity_server_details':
+            return json_payload
+        elif not json_payload.get("count") or json_payload["count"] == 0:
+            check.log.trace("No results found for resource %s url: %s", resource_name, resource_url)
```

Comment:
Curious as to why these are trace log and not debug?

Reply:
The logs can be very spammy if the REST requests succeed but have 0 count of results, such as when there are no new builds for a given build config. I originally had them as debug-level. Do you think the increased verbosity is okay for debug?

Reply:
Hmm that's a good point. The problem that I have with trace is that enabling it is a hassle and that sometimes log rotate out in a short timespan due to logging of metric raw metric values. I personally would prefer debug. But it's a non blocking nit if you feel like trace would be better.